src/app/App.tsx

```diff
@@ -1,59 +1,85 @@
-import { useEffect, useState } from 'react';
+import { useCallback, useEffect, useReducer } from 'react';
 import { getWorkingDirectory } from '../support/nana';
+import { randomId } from '../support/randomId';
+import { Card, CardPropTypes, ICard } from './Card';
 
-import { Card, ICard, CardPropTypes } from './Card';
+type HistoryState = string[];
 
-export default () => {
-  const [history, setHistory] = useState<string[]>([]);
-  const [cards, setCards] = useState<ICard[]>([]);
-
-  useEffect(() => {
-    if (cards.length === 0) addEmptyCard();
-  }, [cards.length]);
+type HistoryAction = {
+  type: 'push';
+  input: string;
+};
 
-  const addEmptyCard = async () => {
-    addCard({
-      workingDir: await getWorkingDirectory(),
-    });
-  };
+const historyReducer = (state: HistoryState, action: HistoryAction) => {
+  switch (action.type) {
+    case 'push':
+      return [...state, action.input];
+  }
+};
 
-  const addCard = (props: CardPropTypes) => {
-    setCards((cards) => [...cards, { id: cards.length, ...props }]);
-  };
+type CardsState = ICard[];
 
-  const addToHistory = (input?: string) => {
-    if (!input) return;
-    if (history.indexOf(input) === -1) {
-      setHistory((history) => [...history, input]);
+type CardsAction =
+  | {
+      type: 'add';
+      card: ICard;
     }
-  };
-
-  const removeCard = (cardId: number) => {
-    setCards((cards) => cards.filter(({ id }) => id !== cardId));
-  };
+  | {
+      type: 'remove';
+      cardId: string;
+    }
+  | {
+      // TODO: break this action into multiple, more explicit, actions
+      type: 'update';
+      cardId: string;
+      props: Partial<CardPropTypes>;
+    };
 
-  const updateCard = (cardId: number, props: CardPropTypes) => {
-    return setCards((cards) =>
-      cards.map((orig) => {
-        if (orig.id === cardId) {
-          return { ...orig, ...props };
+const cardsReducer = (state: CardsState, action: CardsAction) => {
+  switch (action.type) {
+    case 'add':
+      return [...state, action.card];
+    case 'remove':
+      return state.filter((card) => card.id === action.cardId);
+    case 'update':
+      return state.map((orig) => {
+        if (orig.id === action.cardId) {
+          return { ...orig, ...action.props };
         }
         return orig;
-      })
-    );
-  };
+      });
+  }
+};
+
+export const App = () => {
+  const [history, dispatchHistory] = useReducer(historyReducer, []);
+  const [cards, dispatchCards] = useReducer(cardsReducer, []);
+
+  const addEmptyCard = useCallback(async () => {
```

Comment:
My understanding is that [`useCallBack`](https://reactjs.org/docs/hooks-reference.html#usecallback) is a performance optimization; is that correct?

I don't know enough about React to understand why the optimization is important/worthwhile here.

Reply:
That’s right. In this case, I think its not really necessary. Did it out of habit. I’ll remove.

Reply:
Honestly in this case, I think its worse to have it.